Add theme class to body and fix for theme selection

janis91 opened the pull request:
I added a selector for the class ``custom-icon`` in combination with any other ``icon-`` class that can be used by app developers to invert the color of their custom app icon when in dark mode. Especially in positions like the file action menu, where the icon has to be provided as class and not as url for an ``<img>`` tag. 

I think that could be a quick win to support app devs in supporting dark mode.

Commit:
Add theme class for selected theme to body, fix for accessibility theme selection and separated highcontrast theme

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

juliushaertl commented on apps/accessibility/css/themedark.scss:
```diff
@@ -85,3 +85,8 @@ input[type=checkbox] {
+
+.custom-icon[class^='icon-'],
+.custom-icon[class*=' icon-'] {
+    filter: invert(100%);
```
The issue with this is that it doesn't work with IE11

juliushaertl replied:
@skjnldsv and me discussed that we would add a themedark class to the body element so apps can use that to ship two versions of their icon which are then applied depending on the acessibility setting.

janis91 replied:
> The issue with this is that it doesn't work with IE11

I think this applies for all the other invert filters that are already set here, as well.

> @skjnldsv and me discussed that we would add a themedark class to the body element so apps can use that to ship two versions of their icon which are then applied depending on the acessibility setting.
That'd make even more sense. I could do that, too. If you like

janis91 replied:
@juliushaertl @skjnldsv I changed the PR to what you were planning to build (including a fix for a bug I found when I was testing my feature)

skjnldsv replied:
@janis91 thanks :)

Well, all the pages aren't changed with how you did it, right?
Don't you need to have this on the TemplateManager.php so that every page on nextcloud has either the day or night body class?

https://github.com/nextcloud/server/issues/12276#issuecomment-521908405

janis91 replied:
@skjnldsv The accessibility app, if activated, as far as I understand it, loads a css AND a js on every page. What I did, was to add a check on the Controller side, if a theme is activated or not and depending on this, send an additional command with the js file (that adds the corresponding theme class to the page's body).

skjnldsv replied:
Ah yes!

We can indeed  do this in js, as it will at least make sure we're not using accessibility code into core.

skjnldsv commented on apps/accessibility/lib/Controller/AccessibilityController.php:
```diff
@@ -204,13 +204,20 @@ public function getJavascript(): DataDownloadResponse {
 		theme: ' . json_encode($theme) . ',
-		
 	};
+	document.body.classList.add(' . json_encode($theme) . ');
```
I would prefer a day/night class please for the dark theme :)

janis91 replied:
> I would prefer a day/night class please for the dark theme :)

I used the theme id (themedark, themehighcontrast), because we even have more than day/night now already and there might even come more themes in the future (who knows).

skjnldsv replied:
@janis91 that is not planned :)
For sure not in the accessibility app though ;)

And even with other external apps, day/night would still make sense as people could propose dark/white variants of their designs.

janis91 replied:
But isn't the themehighcontrast already one thing that exceeds the 2 options day/night.
We have three options at the moment:

1. "day" / no theme => no class will be added to the body, everything has it's default theme
2. "themedark" => at the moment the id is used as class name, so `themedark` class will be added to the body
3. "themehighcontrast" => at the moment the id is used as class name, so `themehighcontrast` class will be added to the body

So even for themehighcontrast people could now add their adjusments in design. I am not convinced, that an additional mapping like `themedark` => `night` really makes sense. The themedark class would have the same effect as a night class.

skjnldsv replied:
@janis91 because it's cleaner and nicer looking than `themedark`.
And currently we do not support both high contrast and dark, but this is planned. :)
You asked 
> Please let me know if this was what you expected ;-)

Well, we wanted to just have a plain toggle for dark/day them so we can upgrade all of our apps accordingly. There is no plan for apps to have their own custom design for highcontrast.

janis91 replied:
> There is no plan for apps to have their own custom design for highcontrast.

From my point of view there are two points than:
1. Why not rename the theme then? Because then it would be way easier to name it "night" or "dark" directly.
2. Why giving the option of highcontrast, if it's not planned to support it with apps? I mean: If apps won't support it anyway, you could just delete this option (not the best idea from an accessibility point of view).

Don't get me wrong, if `night`or `themedark` is absolutely not the point. It's the software design perspective. Why using `themedark` or `themehighcontrast` as id everywhere else like in the `OCA.Accessibility` object and also internally in the UserConfig, but name it `night` as css class. That can lead to more confusion than a simple class name adds simplicity. When you rename the whole thing to "night", such that the UserConfig key is `theme` and the value is `night` it makes sense to even name the css class the same way. I just want to avoid confusion here.

skjnldsv replied:
> 2\. Why giving the option of highcontrast, if it's not planned to support it with apps? I mean: If apps won't support it anyway, you could just delete this option (not the best idea from an accessibility point of view).

Because the way the high contrast theme is managed is totally independant from the apps. It does not requires any change from the apps to work there :)



> 1. Why not rename the theme then? Because then it would be way easier to name it "night" or "dark" directly.

Because I designed the accessibility themes and apps and I was about to implement the body class toggles for 18 as well. And I think it looks nicer for apps to have a css toggle like
``` scss 
.night {
	.icon-edit {
		xxx: xxxxx
	}
}
```

You asked if that is how I wanted things to be, so I said no because thats not how I wanted things to be. But you then argue how you want things to be :grin: 
Don't ask other's opinions if you don't want to change things then :see_no_evil: 
Nonetheless, it's only my point of view! I don't speak for everyone :wink: 
And to be fair, we're only talking about a css class toggle!

So sure go ahead, just add theme-dark if you want on the body. The goal is to have two classes to differentiate the dark theme from the regular theme. I don't really care as long as it make sense.

skjnldsv replied:
> That can lead to more confusion than a simple class name adds simplicity. When you rename the whole thing to "night", such that the UserConfig key is `theme` and the value is `night` it makes sense to even name the css class the same way.

Let me also add that because I choose a name for the config on the backend does NOT mean we have to use the same config name everywhere on nextcloud. We can do whatever we want anyway, and designers does not care at all what user config key is named like this or like that. :smile_cat:

janis91 replied:
> Because the way the high contrast theme is managed is totally independant from the apps. It does not requires any change from the apps to work there :)

Can you explain that in detail? Because then I might not understand the point of the high contrast theme exactly.

> We can do whatever we want anyway, and designers does not care at all which user config key is named like this or like that.

When the name is used in the `OCA.Accessibility` object - and it is used there - they might get confused, that's all I wanted to say.

> Don't ask other's opinions if you don't want to change things then 🙈

I am open for discussion and I am open to change things, when they make sense. I didn't know that you already implemented changes for apps, that's of course a good point. But why am I supposed to change things just because I ask for opinions? I don't want to convince you with my opinion, I just asked critically, what's the argument behind adding an additional static mapping in the controller code and that it is "only" a css class toggle, that we are talking about, was exactly the point behind my question. Why adding so much in the background just because of another name? I wanted to understand that and not forcing my opinion or something.

I added this feature (and the fix for the bug in the current version) because I wanted to add the dark-mode to my app basically. I wasn't aware, that you already work at this and I just wanted to add it, such that it gets shipped faster and in order to help. It was not my intention to attack your development or your plan. I just wanted to understand this plan.

skjnldsv replied:
> When the name is used in the `OCA.Accessibility` object - and it is used there - they might get confused, that's all I wanted to say.

Ah you're right! I forgot we also advertised it there!
That's a very good point! Let's leave it like you did then, it totally make sense to keep this unified then!

> I added this feature (and the fix for the bug in the current version) because I wanted to add the dark-mode to my app basically. I wasn't aware, that you already work at this and I just wanted to add it, such that it gets shipped faster and in order to help. It was not my intention to attack your development or your plan. I just wanted to understand this plan.

Ahaha, don't worry :)
I was just very confused because you seemed more determined to do it your way than mine, which confused me on why you asked for how we wanted to do things here! Of course your opinion is as valuable as mine or anyone else ;)

> Can you explain that in detail? Because then I might not understand the point of the high contrast theme exactly.

We're applying a css override on each page to remove any opacity or slightly lowered contrast values. It does not matter if the app changed something, we will show a text pure black even if it was set to slight grey. :wink: 



> Why adding so much in the background just because of another name? I wanted to understand that and not forcing my opinion or something.

What do you mean by "addign so much in the background?" ? :)

skjnldsv replied:
> When the name is used in the `OCA.Accessibility` object - and it is used there - they might get confused, that's all I wanted to say.

Ah you're right! I forgot we also advertised it there!
That's a very good point! Let's leave it like you did then, it totally make sense to keep this unified then!

> I added this feature (and the fix for the bug in the current version) because I wanted to add the dark-mode to my app basically. I wasn't aware, that you already work at this and I just wanted to add it, such that it gets shipped faster and in order to help. It was not my intention to attack your development or your plan. I just wanted to understand this plan.

Ahaha, don't worry :)
I was just very confused because you seemed more determined to do it your way than mine, which confused me on why you asked for how we wanted to do things here! Of course your opinion is as valuable as mine or anyone else ;)

> Can you explain that in detail? Because then I might not understand the point of the high contrast theme exactly.

We're applying a css override on each page to remove any opacity or slightly lowered contrast values. It does not matter if the app changed something, we will show a text pure black even if it was set to slight grey. :wink: 



> Why adding so much in the background just because of another name? I wanted to understand that and not forcing my opinion or something.

What do you mean by "addign so much in the background?" ? :)

janis91 replied:
> What do you mean by "adding so much in the background?" ? :)

I meant adding an additional static mapping from the theme's id to the css class.

> It does not matter if the app changed something, we will show a text pure black even if it was set to slight grey.

Ok then I understand the idea behind the high contrast topic. But if I look at my app for example, where I have a dialog/modal which adds also a hint in grey text for example. Shouldn't I change this to black when the highcontrast theme is activated, as well? Because  how do you overwrite this for my app then?

skjnldsv replied:
> But if I look at my app for example, where I have a dialog/modal which adds also a hint in grey text for example.

Could you provide an example for me? :)

> I meant adding an additional static mapping from the theme's id to the css class.

Ah yes sure! Then we need to think about the future. The way we store the themedark and stuff is not appropriate, because it does not allow for multiple theme. Maybe just storing dark/highcontrast is better here.

janis91 replied:
> Could you provide an example for me? :)

https://github.com/janis91/ocr/blob/b0afbbb37efe5a18d39ef7077425d20c1f8662e7/css/app.css#L198
Here I added a class for a hint that gets displayed in my app. The color could also be changed depending on the `themehighcontrast`  or `highcontrast` toggle class at the body tag. But I cannot see a possibility to change this globally by overriding something. So I thought, that this might not be the only example. There may be other apps where this applies as well and a custom class toggle makes sense.

> because it does not allow for multiple theme

Ah then my "fix" is also not right here. Because if it is allowed to activate this at the same time, I stoped it from working. But as you mentioned already, then we would need one more key for high contrast, that's right.

janis91 replied:
Of course I could have used the css variables for light text color here. But there might be other examples, where this is not fitting.

skjnldsv replied:
@janis91 actually yes, you should use the css variables for your app to be compliant with nextcloud :)
https://github.com/nextcloud/server/blob/679a694e6f7fc77551a21f106d382b5f1800c6bb/core/css/css-variables.scss#L25

janis91 replied:
Yes I know, that's what I will do when my dark-mode changes get integrated.
But: I can also think of other apps using different colors, like a light green or light red. That would cannot be change globally. So I think it might still be worth it, to provide a `highcontrast` toggle as well. Even though for the core and most apps it is not necessary. But at this point I am not sure. ;-)

skjnldsv replied:
> But: I can also think of other apps using different colors, like a light green or light red. That would cannot be change globally

We have those as well :)
To be fair, you should not use custom colours in nextcloud apps.
We provide everything you need. If you starts using other things, you will just break the design unity we try so hard to have on all of our apps ;)

If it would be me, I would display errors to every dev that does not use css variables :see_no_evil:

janis91 replied:
> To be fair, you should not use custom colours in nextcloud apps.

That's fair. It was just an idea.

So to sum up:
I will

1. add another user config entry/key for high contrast theme
2. rename the `themedark` and `themehighcontrast` id. the question is, if to `dark`  and `highcontrast` OR if we introduce another user config entry anyway, could just save a boolean for both: dark and high contrast theme. 
3. Change the behavior of the vue app again, such that high contrast AND dark theme are possible at the same time
4. Only the `dark` css toggle will get added with JS if active, no class like `day` or similiar will be set AND the highcontrast theme will not get a css toggle, either.

What do you think especially with regards to point 2?

skjnldsv replied:
That would we awesome :)
Thanks for taking care of this!!!

For 2 I would say maybe add a dedicated entry for highcontrast? :thinking:

janis91 replied:
@skjnldsv done. You can review now :-) 
I know the underlying json structure to separate highcontrast is not really pretty, but I could leave most of the vue app code untouched, because I used the same structure as for "themes".

skjnldsv commented on apps/accessibility/src/App.vue:
```diff
@@ -81,8 +88,13 @@ export default {
+		selectTheme(id, idSelectedBefore) {
 			this.selectItem('theme', id);
+			document.body.classList.remove(idSelectedBefore);
+			if (id) document.body.classList.add(id);
```
```suggestion
			if (id) {
				document.body.classList.add(id);
			}
```

skjnldsv reviewed (changes requested):
We should also keep the old values.
Maybe a repair step to make sure that all old user config is migrated to the new name? :)

janis91 commented:
> We should also keep the old values.
Maybe a repair step to make sure that all old user config is migrated to the new name? :)

Good point. I already thought about that, but forgot it when I wrote down the summary ;-)
I will update soon.

Commit:
Added migration step

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

janis91 commented:
I updated the PR but I always get a weird Autoload error:
``\OC\Repair::error: Repair error: Autoload path not allowed: /***/***/***/nextcloud/apps/accessibility/lib/migration/repairuserconfig.php``
Do you know what's the point here? Because for my app I did the same setup and it works there.
Why is this path not allowed to autoload?

skjnldsv commented:
Did you run the update autoload script in the build directory?

janis91 commented:
> Did you run the update autoload script in the build directory?

Sorry, what do you mean by that?

I just added the Migration/Repair step and updated the app's version. And when NC tries to run it, it fails with that message.

skjnldsv committed:
Update autoload checkers

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
> Sorry, what do you mean by that?

https://drone.nextcloud.com/nextcloud/server/21319/3/3
```
The autoloaders are not up to date
Please run: bash build/autoloaderchecker.sh
And commit the result
```

skjnldsv commented on .htaccess:
```diff
@@ -66,3 +66,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
```
To remove :)

skjnldsv commented on apps/accessibility/src/components/itemPreview.vue:
```diff
@@ -14,18 +14,19 @@
+				return this.selected === this.preview.id;
+			},
+			set(val) {
+				this.$emit(
```
Very clever! :)
Though could you put the syntax inline, this is quite hard to read! :see_no_evil: 

(we still haven't enabled our eslint config on server, all other apps are, but not here yet ^^)

kesselb commented on apps/accessibility/lib/AccessibilityProvider.php:
```diff
@@ -56,19 +56,25 @@ public function __construct(string $appName,
+			[
+				'id'    => 'highcontrast',
+				'img'   => $this->urlGenerator->imagePath($this->appName, 'theme-highcontrast.jpg'),
+				'title' => $this->l->t('High contrast theme'),
```
Is it still a theme?

skjnldsv replied:
We can still calling it that yes. Did you have anything else in mind?

janis91 replied:
One related issue at the moment is that if both themes are activated, it's not dark anymore. So I guess we have to think about that to. @skjnldsv What was your idea here? Should we add another highcontrast dark scss here and import that instead, when dark theme is activated?

kesselb replied:
> Did you have anything else in mind?

No :man_shrugging: High contrast theme is independent then. So high contrast is more a filter or a flag? Usually there is only one theme.

skjnldsv replied:
Yeah, it's more like a setting.
We need to adapt the script to load the two css at the same time :)

janis91 replied:
> High contrast theme is independent then. So high contrast is more a filter or a flag? Usually there is only one theme.

@kesselb I think I get what you mean: The highcontrast getter should return an object instead of an array actually. I changed this now.

kesselb replied:
My comment was about the word "theme" :confused:  But if you think an object is better here fine by me :+1:

janis91 replied:
@kesselb @skjnldsv Is there a better word than "theme" for the highcontrast topic. I can only think of "mode" or "style". I added the high contrast styling part for dark mode already (now there is one for the day/light theme and for the night/dark).

skjnldsv replied:
`High contrast mode` ?

Commit:
Add small linting fix

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

Commit:
Changed highcontrast from array (theme) to object (flag/filter)

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

Commit:
Add highcontrast style for dark mode

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

Commit:
Change name of high contrast theme to high contrast mode

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

kesselb commented on apps/accessibility/lib/Migration/RepairUserConfig.php:
```diff
@@ -0,0 +1,89 @@
+declare (strict_types = 1);
+/**
+ * @author Janis Koehr <janiskoehr@icloud.com>
+ * @copyright Janis Koehr 2019
```
https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md

It probably doesn't matter but the copyright usually looks a bit different.

janis91 replied:
Added in the right format

kesselb commented on apps/accessibility/lib/Controller/AccessibilityController.php:
```diff
@@ -132,6 +132,9 @@ public function getCss(): DataDisplayResponse {
 		foreach ($userValues as $key => $scssFile) {
 			if ($scssFile !== false) {
+				if ($scssFile === 'highcontrast' && in_array('dark', $userValues)) {
+					$scssFile .= 'dark';
```
If we add more themes here we may think about some convention like `highcontrast.$themeName`. But something for later because we need to kill the loop then.

janis91 replied:
Yes that's right. The loop makes no more sense then.

kesselb reviewed (commented):
Thanks for your contribution :+1:

Commit:
Added copyright for contribution

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

janis91 commented:
> Thanks for your contribution 👍

@kesselb something else I should add?

kesselb commented:
> something else I should add?

cc @juliushaertl @ChristophWurst

juliushaertl commented on apps/accessibility/lib/Migration/RepairUserConfig.php:
```diff
@@ -0,0 +1,88 @@
+class RepairUserConfig implements IRepairStep {
+
+	/** @var string */
+	private const APP_NAME = 'accessibility';
```
Please use Application::APP_NAME, no need to define this twice

janis91 replied:
@juliushaertl good point. should be resolved now.

juliusknorr reviewed (commented):
Looks good besides the small comment and works as advertised.

juliushaertl commented:
Also please rebase and rebuild the bundles

Commit:
Removed duplicate defintion of app name and rebased/rebuilt.

Signed-off-by: Janis Köhr <janis.koehr@novatec-gmbh.de>

janis91 commented:
> Also please rebase and rebuild the bundles

done.
